### PR TITLE
IExceptionHandler for exception handler middleware #46280

### LINKED ISSUE #46280: IExceptionHandler for exception handler middleware
## Background and Motivation

The exception handling middleware is designed to take unexpected, unhandled exceptions, log them, and render a customizable response for the client. This is a last ditch effort to produce something human readable, otherwise the response will receive an empty 500. This is not intended for known cases or control flow. Expected exceptions should be handled at the individual call sites.

Ask: Some exceptions like database exceptions could happen in many places and having to handle those in each place requires a lot of redundant code. How can handling of known exceptions be centralized?

## Proposed API

The IExceptionHandler is a new interface that can have multiple instances registered with DI. The existing exception handling middleware will resolve these and loop through them until something matches. If none match, a higher level log (warning?) will be written. Exception handlers are not run if the response has already started. The existing IDeveloperPageExceptionFilter is used similarly with the developer exception page.

```diff
Assembly: Microsoft.AspNetCore.Diagnostics.Abstractions

Microsoft.Extensions.DependencyInjection;

public static class ExceptionHandlerServiceCollectionExtensions
{
+  public static IServiceCollection AddExceptionHandler<T>(this IServiceCollection services) where T is IExceptionHandler;
}

namespace Microsoft.AspNetCore.Diagnostics;

+ public interface IExceptionHandler
+ {
+    ValueTask<bool> TryHandleAsync(HttpContext httpContext, Exception exception, CancellationToken cancellationToken);
+ }

```

## Usage Examples

```csharp

services.AddExceptionHandler<DBExceptionHandler>();

public class DBExceptionHandler : IExceptionHandler
{
  ValueTask<bool> TryHandleAsync(HttpContext httpContext, Exception exception, CancellationToken cancellationToken)
  {
    if (exception is DBException) ...
  };
}
```

## Risks

We don't want to encourage developers to use exceptions as normal control flow, it's extremely in-efficient and poor design. We should document that expectation and avoid using this capability in our own components.

This should not be confused with the MVC IExceptionFilter.
https://learn.microsoft.com/en-us/dotnet/api/system.web.mvc.iexceptionfilter

### PR DESCRIPTION
**Pull request by Kahbazi**
Fixes https://github.com/dotnet/aspnetcore/issues/46280

### CODE REVIEW

**Review by Tratcher — APPROVED** [review]

**Comment by davidfowl** [comment]
@Tratcher is there a reason you didn't merge this?

File: src/Middleware/Diagnostics/src/ExceptionHandler/ExceptionHandlerMiddlewareImpl.cs [hunk 1]
```diff
@@ -34,19 +35,22 @@ internal class ExceptionHandlerMiddlewareImpl
...
         DiagnosticListener diagnosticListener,
+        IEnumerable<IExceptionHandler> exceptionHandlers,
         IProblemDetailsService? problemDetailsService = null)
     {
         _next = next;
         _options = options.Value;
         _logger = loggerFactory.CreateLogger<ExceptionHandlerMiddleware>();
         _clearCacheHeadersDelegate = ClearCacheHeaders;
         _diagnosticListener = diagnosticListener;
+        _exceptionHandlers = exceptionHandlers;
```
Comment (davidfowl) [inline src/Middleware/Diagnostics/src/ExceptionHandler/ExceptionHandlerMiddlewareImpl.cs:53]: Turn this into a list or array so we can avoid the extra enumerator allocation on exceptions.

Comment (davidfowl) [reply src/Middleware/Diagnostics/src/ExceptionHandler/ExceptionHandlerMiddlewareImpl.cs:53]: See https://github.com/dotnet/runtime/blob/953d290e756d29707b9829e3ea00ef8f2fc31a9f/src/libraries/Microsoft.Extensions.Options/src/OptionsFactory.cs#L38-L45

**Comment by davidfowl** [comment]
Do we have any plans to use this outside of the unit tests?

File: src/Middleware/Diagnostics/src/ExceptionHandler/ExceptionHandlerMiddlewareImpl.cs (hunk 1 above)
Comment (Kahbazi) [reply src/Middleware/Diagnostics/src/ExceptionHandler/ExceptionHandlerMiddlewareImpl.cs]: Done!

**Comment by davidfowl** [comment]
> We haven't identified anywhere in the product we'd use this feature. I suppose we could refactor the BadHttpRequest and ProblemDetails handling to use it?

Yes, I was hoping this would be the case.

**Comment by davidfowl** [comment]
> We haven't identified anywhere in the product we'd use this feature. I suppose we could refactor the BadHttpRequest and ProblemDetails handling to use it?

Can we track this @Tratcher ?

**Pull request merged**